[reland][quant] Remove nn.quantized.ReLU module and nn.quantized.functional.relu (#47415)

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48059 [quant][fix] Fix quantization for qat.ConvBnReLU1d
* **#48038 [reland][quant] Remove nn.quantized.ReLU module and nn.quantized.functional.relu (#47415)**

Summary:
nn.ReLU works for both float and quantized input, we don't want to define an nn.quantized.ReLU
that does the same thing as nn.ReLU, similarly for nn.quantized.functional.relu

this also removes the numerical inconsistency for models quantizes nn.ReLU independently in qat mode

Test Plan: Imported from OSS

Reviewed By: z-a-f

Differential Revision: [D25000462](https://our.internmc.facebook.com/intern/diff/D25000462)

jerryzh168 committed:
[reland][quant] Remove nn.quantized.ReLU module and nn.quantized.functional.relu (#47415)

Summary:
nn.ReLU works for both float and quantized input, we don't want to define an nn.quantized.ReLU
that does the same thing as nn.ReLU, similarly for nn.quantized.functional.relu

this also removes the numerical inconsistency for models quantizes nn.ReLU independently in qat mode

Test Plan: Imported from OSS

Reviewed By: z-a-f

[ghstack-poisoned]

jerryzh168 committed:
Update on "[reland][quant] Remove nn.quantized.ReLU module and nn.quantized.functional.relu (#47415)"


Summary:
nn.ReLU works for both float and quantized input, we don't want to define an nn.quantized.ReLU
that does the same thing as nn.ReLU, similarly for nn.quantized.functional.relu

this also removes the numerical inconsistency for models quantizes nn.ReLU independently in qat mode

Test Plan: Imported from OSS

Reviewed By: z-a-f

Differential Revision: [D25000462](https://our.internmc.facebook.com/intern/diff/D25000462)

[ghstack-poisoned]

vkuzo commented on torch/nn/intrinsic/quantized/modules/conv_relu.py:
```diff
@@ -39,6 +39,10 @@ def _get_name(self):
 
     @classmethod
     def from_float(cls, mod):
+        if type(mod) == torch.nn.intrinsic.qat.ConvBnReLU1d:
+            mod.weight, mod.bias = fuse_conv_bn_weights(
+                mod.weight, mod.bias, mod.bn.running_mean, mod.bn.running_var,
+                mod.bn.eps, mod.bn.weight, mod.bn.bias)
```
thoughts about moving this to a separate PR to minimize risk?

jerryzh168 replied:
sure

vkuzo commented on torch/nn/quantized/modules/batchnorm.py:
```diff
@@ -21,11 +21,9 @@ def _get_name(self):
 
     @classmethod
     def from_float(cls, mod):
+        activation_post_process = mod.activation_post_process
```
any chance you could expand on the reasoning for this change?

jerryzh168 replied:
sure, previous code is relying on the fact the we attach activation_post_process to ReLU inside the sequential, but this PR removes activation_post_process for ReLU, so we changed the insert observer logic to add activation_post_process to sequential directly so we don't need to get activation_post_process from relu any more
```
activation_post_process = mod[1].activation_post_process	
```
this is getting activation_post_proceess from relu.

vkuzo replied:
makes sense, thanks!  maybe we can add a quick comment to the code?

jerryzh168 replied:
will add a comment to the place we attach activation_post_process for _FusedModule (in quantization/quantize.py)

vkuzo commented on torch/quantization/quantize.py:
```diff
@@ -137,14 +136,18 @@ def insert_activation_post_process(m, special_act_post_process=None):
             m._forward_hooks.move_to_end(handle.id, last=False)
 
     for name, child in module.named_children():
```
would it make sense to add a stronger test for this section to help catch future issues?  For example, something like a model which has modules going through all of the branches in these if statements, passing some data through it, and verifying that every observer has seen values?

jerryzh168 replied:
I think we do have tests for each of the branches, although it's probably spread out in different op tests.

jerryzh168 replied:
if we make an artificial model, not sure it would be runnable or not actually.

vkuzo replied:
it doesn't have to be this PR (since we are adding an fb-only test), but seems like just increasing integration model coverage for matching numerics between eager and FX would be valuable.  It seems like `TestQuantizeFxModels` has various models commented out, perhaps we can invest in increasing coverage a bit.

jerryzh168 replied:
yeah I think we should add more e2e tests as well

jerryzh168 committed:
Update on "[reland][quant] Remove nn.quantized.ReLU module and nn.quantized.functional.relu (#47415)"


Summary:
nn.ReLU works for both float and quantized input, we don't want to define an nn.quantized.ReLU
that does the same thing as nn.ReLU, similarly for nn.quantized.functional.relu

this also removes the numerical inconsistency for models quantizes nn.ReLU independently in qat mode

Test Plan: Imported from OSS

Reviewed By: z-a-f

Differential Revision: [D25000462](https://our.internmc.facebook.com/intern/diff/D25000462)

[ghstack-poisoned]

vkuzo reviewed (approved):
lg if the fb-only test passes, thanks!

jerryzh168 committed:
Update on "[reland][quant] Remove nn.quantized.ReLU module and nn.quantized.functional.relu (#47415)"


Summary:
nn.ReLU works for both float and quantized input, we don't want to define an nn.quantized.ReLU
that does the same thing as nn.ReLU, similarly for nn.quantized.functional.relu

this also removes the numerical inconsistency for models quantizes nn.ReLU independently in qat mode

Test Plan: Imported from OSS

Reviewed By: z-a-f

Differential Revision: [D25000462](https://our.internmc.facebook.com/intern/diff/D25000462)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[reland][quant] Remove nn.quantized.ReLU module and nn.quantized.functional.relu (#47415)"


Summary:
nn.ReLU works for both float and quantized input, we don't want to define an nn.quantized.ReLU
that does the same thing as nn.ReLU, similarly for nn.quantized.functional.relu

this also removes the numerical inconsistency for models quantizes nn.ReLU independently in qat mode

Test Plan: Imported from OSS

Reviewed By: z-a-f

Differential Revision: [D25000462](https://our.internmc.facebook.com/intern/diff/D25000462)

[ghstack-poisoned]
